File: problems/0002-add-two-numbers/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head_ptr = new ListNode();
        ListNode* tail_ptr = head_ptr;
        int carry = 0;

        do {

            int l1_val = l1 == nullptr ? 0 : l1->val;
            int l2_val = l2 == nullptr ? 0 : l2->val;

            int sum = l1_val + l2_val + carry;
            tail_ptr->val = sum % 10;
            carry = sum > 9 ? 1 : 0;

            bool l1_has_next = (l1 != nullptr && l1->next != nullptr);
            bool l2_has_next = (l2 != nullptr && l2->next != nullptr);

            if (l1_has_next || l2_has_next || carry > 0) {
                tail_ptr->next = new ListNode();
                tail_ptr = tail_ptr->next;

                l1 = l1 == nullptr ? nullptr : l1->next;
                l2 = l2 == nullptr ? nullptr : l2->next;
            } else {
                return head_ptr;
            }

        } while (true);
    }
};

File: problems/0002-add-two-numbers/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dhead = ListNode()
        curr = dhead
        carry = 0

        while l1 or l2 or carry:
            l1val = l1.val if l1 else 0
            l2val = l2.val if l2 else 0
            total = l1val + l2val + carry

            carry = total // 10
            total = total % 10

            curr.next = ListNode(total)
            curr = curr.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        
        self.printList(dhead.next)
        return dhead.next

    def printList(self, head):
        s = ""
        while head:
            s += f"{head.val} -->"
            head = head.next
        
        print(s)
